pi5/pinctrl_utils.go

```diff
@@ -0,0 +1,231 @@
+			childAddr = uint64(binary.BigEndian.Uint32(rangeByteContents[:(4 * numCAddrCells)]))
+		} else if numCAddrCells == 2 { // reading in 64 bits
+			childAddr = binary.BigEndian.Uint64(rangeByteContents[:(4 * numCAddrCells)])
+		} else { // reading in more than 64 bits. we only want the last 64 bits of the address though, so we cut off the other portion of it
+			childAddr = binary.BigEndian.Uint64(rangeByteContents[(4 * (numCAddrCells - 2)):(4 * numCAddrCells)])
+		}
+		rangeByteContents = rangeByteContents[(4 * numCAddrCells):] // flush the bytes already parsed out of the array
+
+		if numPAddrCells == 1 { // reading in 32 bits. regardless we must convert to a 64 bit address so we add a bunch of 0s to the beginning.
+			parentAddr = uint64(binary.BigEndian.Uint32(rangeByteContents[:(4 * numPAddrCells)]))
+		} else if numPAddrCells == 2 { // reading in 64 bits
+			parentAddr = binary.BigEndian.Uint64(rangeByteContents[:(4 * numPAddrCells)])
+		} else { // reading in more than 64 bits. we only want the last 64 bits of the address though, so we cut off the other portion of it
+			parentAddr = binary.BigEndian.Uint64(rangeByteContents[(4 * (numPAddrCells - 2)):(4 * numPAddrCells)])
+		}
+		rangeByteContents = rangeByteContents[(4 * numPAddrCells):]
+
+		if numAddrSpaceCells == 1 { // reading in 32 bits. regardless we must convert to a 64 bit address so we add a bunch of 0s to the beginning.
+			addrSpaceSize = uint64(binary.BigEndian.Uint32(rangeByteContents[:(4 * numAddrSpaceCells)]))
+		} else if numAddrSpaceCells == 2 { // reading in 64 bits
+			addrSpaceSize = binary.BigEndian.Uint64(rangeByteContents[:(4 * numAddrSpaceCells)])
+		} else { // reading in more than 64 bits. we only want the last 64 bits of the address though, so we cut off the other portion of it
+			addrSpaceSize = binary.BigEndian.Uint64(rangeByteContents[(4 * (numAddrSpaceCells - 2)):(4 * numAddrSpaceCells)])
+		}
+
+		rangeByteContents = rangeByteContents[(4 * numAddrSpaceCells):]
+		rangeInfo := rangeInfo{childAddr: childAddr, parentAddr: parentAddr, addrSpaceSize: addrSpaceSize}
+		addrRangesSlice = append(addrRangesSlice, rangeInfo)
+
+	}
+
+	return addrRangesSlice, err
+}
+
+// Uses Information Stored within the 'reg' property of the child node and 'ranges' property of its parents to map the child's physical address into the dev/gpiomem space
+func (b *pinctrlpi5) setGPIONodePhysAddr(nodePath string) error {
+
+	currNodePath := dtBaseNodePath + nodePath // initially: /proc/device-tree/axi/pcie@120000/rp1/gpio@d0000
+	var numCAddrCells uint32 = 0
+	var err error = nil
+
+	/* Call Recursive Function to Calculate Phys Addr. Works way up the Device Tree, using the information
+	found in #ranges at every node to translate from the child's address space to the parent's address space
+	until we get the child's physical address in all of /dev/gpiomem. */
+	*b.physAddr, err = setGPIONodePhysAddrHelper(currNodePath, INVALID_ADDR, numCAddrCells)
```

Comment:
can you use 0 instead of invalid_addr here?

Reply:
I didn't want to use 0 at all in my physical address translation because technically 0 is within the range of the address space, and the offset could technically be 0 and it would create a weird edge case. I wanted to have INVALID_ADDR be passed in because its a max integer, which is basically guaranteed to be outside the range of the address space. does that make sense?

Reply:
okay that makes sense for not using zero, what are your thoughts about using `NaN` as the invalid case?

Reply:
that's better for error handling, I'll edit the Invalid_addr parameter